File: Sources/RobotLink/MotionGuard.swift
import Foundation
import RockyKit
import Telemetry

/// Single chokepoint for every motion command Rocky issues to the daemon.
///
/// Every `setTarget`, `goto`, and `playRecordedMove` call MUST go through
/// this guard rather than calling `RobotLinkClient` directly. The guard
/// enforces:
///
/// 1. **Slew-rate limit on `setTarget`** — caps the per-tick change
///    in each joint's commanded target so abrupt jumps in `latest`
///    (face tracker pivoting between subjects, look-at-object firing
///    while the streamer's still pushing the previous pose, etc.)
///    don't translate into a hardware-limited motor slew that reads
///    as a jerk.
///
/// 2. **Velocity clamp on `goto`** — for *all* joints (head, body,
///    antennas). The old `safeGotoDuration` only considered head
///    pose. A big body-yaw delta with a short requested duration
///    could exceed the safe velocity ceiling silently.
///
/// 3. **Duration floor on `goto`** — every goto gets at least
///    `Config.minGotoDurationS` (default 0.4 s). Prevents a
///    callsite from issuing a 50 ms "snap" move.
///
/// 4. **Single-in-flight `goto`** — overlapping gotos cancel each
///    other on the daemon side and produce visible jerks at the
///    transition. The guard serialises gotos so the second waits
///    for the first.
///
/// 5. **Shelf-safe recorded-move allowlist** — `playRecordedMove`
///    only accepts moves in `Config.shelfSafeMoves` unless the
///    explicit `force: true` flag is set. The Pollen emotions
///    library has moves authored for floor-mounted bots that have
///    knocked Rocky off the desk shelf; this gate makes that
///    impossible by default.
///
/// The guard is an actor so multiple concurrent callers serialise
/// through its state (`lastSentTarget`, `inflightGoto`).
public actor MotionGuard {
    public struct Config: Sendable {
        /// Max change in any commanded joint per `setTarget` call.
        /// At 50 Hz, 0.05 rad ≈ 2.9°/tick ≈ 143°/s. Below the
        /// damper's max output speed (1.2 rad/s ≈ 69°/s), so smooth
        /// damped inputs pass through unchanged; sudden jumps get
        /// flattened into a slew.
        public var maxSetTargetSlewRad: Double = 0.05

        /// Goto duration floor. 0.4 s reads as "deliberate but
        /// responsive"; anything shorter snaps and looks frightening.
        public var minGotoDurationS: TimeInterval = 0.4

        /// Max joint velocity for goto duration stretching. Same as
        /// the existing global ceiling so behaviour is consistent.
        public var maxGotoVelocityRadPerS: Double = 1.5

        /// Max difference between head_yaw and body_yaw, in radians.
        /// Pollen documents this as a HARDWARE safety constraint
        /// (anti-collision between the head and body) at 65°. The
        /// daemon clamps to "the closest valid position" silently
        /// when you exceed it — meaning the actual head or body
        /// ends up somewhere other than commanded, which the brain's
        /// next round won't be expecting. Enforcing it here makes
        /// the failure explicit: the goto either targets a feasible
        /// pose or it gets reshaped before send.
        public var maxHeadBodyYawDeltaRad: Double = 65.0 * .pi / 180.0

        /// Recorded moves that have been audited as safe on a
        /// desk-shelf mount (slow, low head excursion, no body
        /// rotation extremes). `playRecordedMove` rejects anything
        /// outside this set unless `force: true`.
        ///
        /// The dance/rage moves are deliberately excluded — they
        /// were the source of the 2026-05-13 shelf incident.
        public var shelfSafeMoves: Set<String> = [
            "amazed1", "attentive1", "calming1", "cheerful1",
            "curious1", "grateful1", "helpful1", "indifferent1",
            "inquiring1", "no1", "no_sad1", "proud1", "relief1",
            "sad1", "serenity1", "shy1", "thoughtful1", "tired1",
            "understanding1", "welcoming1", "yes1", "yes_sad1",
            "downcast1", "lonely1", "loving1", "uncertain1",
        ]

        public init() {}
    }

    private let client: RobotLinkClient
    private let logBus: LogBus
    private(set) public var config: Config

    private var lastSentTarget: MotionTarget?
    private var inflightGoto: Task<Void, Error>?

    public init(
        client: RobotLinkClient,
        logBus: LogBus,
        config: Config = .init()
    ) {
        self.client = client
        self.logBus = logBus
        self.config = config
    }

    public func setConfig(_ config: Config) { self.config = config }

    // MARK: - setTarget (slew-rate limited)

    public func setTarget(_ target: MotionTarget) async throws {
        let limited = slewLimit(target, against: lastSentTarget)
        lastSentTarget = limited
        try await client.setTarget(limited)
    }

    private func slewLimit(
        _ target: MotionTarget, against prev: MotionTarget?
    ) -> MotionTarget {
        guard let prev else { return target }
        let maxSlew = config.maxSetTargetSlewRad
        var limited = target
        if let newHead = target.headPose, let prevHead = prev.headPose {
            limited.headPose = RPYPose(
                roll: clampDelta(newHead.roll, prev: prevHead.roll, maxDelta: maxSlew),
                pitch: clampDelta(newHead.pitch, prev: prevHead.pitch, maxDelta: maxSlew),
                yaw: clampDelta(newHead.yaw, prev: prevHead.yaw, maxDelta: maxSlew)
            )
        }
        if let newAnt = target.antennas, let prevAnt = prev.antennas {
            limited.antennas = Antennas(
                rightRad: clampDelta(newAnt.right, prev: prevAnt.right, maxDelta: maxSlew),
                leftRad: clampDelta(newAnt.left, prev: prevAnt.left, maxDelta: maxSlew)
            )
        }
        if let newBody = target.bodyYaw, let prevBody = prev.bodyYaw {
            limited.bodyYaw = clampDelta(newBody, prev: prevBody, maxDelta: maxSlew)
        }
        return limited
    }

    private func clampDelta(_ new: Double, prev: Double, maxDelta: Double) -> Double {
        let delta = new - prev
        if abs(delta) <= maxDelta { return new }
        return prev + (delta > 0 ? maxDelta : -maxDelta)
    }

    // MARK: - goto (velocity + duration floor + single-in-flight)

    public func goto(
        headPose: RPYPose? = nil,
        antennas: Antennas? = nil,
        bodyYaw: Double? = nil,
        durationS: TimeInterval,
        interpolation: RobotLinkClient.Interpolation = .minjerk
    ) async throws {
        // Serialise with any in-flight goto so a new caller can't
        // interrupt a slow move mid-flight.
        if let pending = inflightGoto {
            _ = try? await pending.value
        }

        // Floor the requested duration.
        let floored = max(durationS, config.minGotoDurationS)

        // **Head-body yaw delta gate.** Pollen's daemon enforces a
        // 65° max between head and body yaw (anti-collision). If the
        // requested pose would exceed that, the daemon silently
        // clamps — leaving the bot somewhere other than commanded.
        // Pull the head and/or body in here so we send a feasible
        // pose that lands where the brain thinks it should.
        let reshapedHead: RPYPose?
        let reshapedBody: Double?
        (reshapedHead, reshapedBody) = await reshapeForYawDelta(
            headPose: headPose, bodyYaw: bodyYaw
        )

        // Compute a safe duration that respects velocity across ALL
        // joints (not just head, which is all RobotLinkClient.goto
        // currently checks).
        let safe = await safeDuration(
            headPose: reshapedHead, antennas: antennas, bodyYaw: reshapedBody,
            requested: floored
        )
        if safe > floored {
            await logBus.publish(.sidecarLog(
                sidecar: "motion-guard",
                level: .info,
                message: String(
                    format: "goto duration stretched %.2fs → %.2fs (velocity cap)",
                    floored, safe
                ),
                fields: [:]
            ))
        }

        let task = Task { [client] in
            try await client.goto(
                headPose: reshapedHead, antennas: antennas, bodyYaw: reshapedBody,
                durationS: safe, interpolation: interpolation
            )
        }
        inflightGoto = task
        defer { inflightGoto = nil }

        try await task.value

        // Sync lastSentTarget so subsequent setTarget slew limiting
        // starts from where the goto actually parked the bot. Use
        // the reshaped poses (post yaw-delta enforcement) since that
        // is what the daemon actually received.
        if let h = reshapedHead {
            lastSentTarget = MotionTarget(
                headPose: h, antennas: antennas ?? lastSentTarget?.antennas,
                bodyYaw: reshapedBody ?? lastSentTarget?.bodyYaw
            )
        }
    }

    /// Enforces Pollen's 65° head-body yaw delta limit. If the requested
    /// (or current-if-not-requested) head and body yaws exceed that
    /// difference, pulls them together so the difference equals the
    /// cap. Splits the correction 50/50 across the two joints; if only
    /// one of head/body is being commanded, the other stays put and
    /// the commanded one absorbs the full correction.
    private func reshapeForYawDelta(
        headPose: RPYPose?, bodyYaw: Double?
    ) async -> (RPYPose?, Double?) {
        // Need both current values to reason about the delta.
        let current: RobotState? = try? await client.fullState()
        let curHeadYaw = current?.headPose.yaw ?? 0
        let curBodyYaw = current?.bodyYaw ?? 0

        let targetHeadYaw = headPose?.yaw ?? curHeadYaw
        let targetBodyYaw = bodyYaw ?? curBodyYaw
        let delta = targetHeadYaw - targetBodyYaw
        let maxDelta = config.maxHeadBodyYawDeltaRad

        if abs(delta) <= maxDelta {
            return (headPose, bodyYaw)
        }

        // Exceeded — pull both toward each other so the new delta
        // equals the cap. If only head OR only body was commanded,
        // the un-commanded one stays at its current value.
        let excess = abs(delta) - maxDelta
        let direction: Double = delta > 0 ? 1 : -1
        var newHeadYaw = targetHeadYaw
        var newBodyYaw = targetBodyYaw
        if headPose != nil && bodyYaw != nil {
            newHeadYaw -= direction * excess / 2
            newBodyYaw += direction * excess / 2
        } else if headPose != nil {
            newHeadYaw -= direction * excess
        } else if bodyYaw != nil {
            newBodyYaw += direction * excess
        }

        await logBus.publish(.sidecarLog(
            sidecar: "motion-guard",
            level: .warn,
            message: String(
                format: "yaw-delta limit: head %.1f° + body %.1f° (Δ %.1f°) → head %.1f° + body %.1f° (Δ %.1f°)",
                targetHeadYaw * 180 / .pi, targetBodyYaw * 180 / .pi,
                delta * 180 / .pi,
                newHeadYaw * 180 / .pi, newBodyYaw * 180 / .pi,
                (newHeadYaw - newBodyYaw) * 180 / .pi
            ),
            fields: [:]
        ))

        let reshapedHead: RPYPose? = headPose.map {
            RPYPose(roll: $0.roll, pitch: $0.pitch, yaw: newHeadYaw)
        }
        let reshapedBody: Double? = bodyYaw.map { _ in newBodyYaw }
        return (reshapedHead, reshapedBody)
    }

    private func safeDuration(
        headPose: RPYPose?, antennas: Antennas?, bodyYaw: Double?,
        requested: TimeInterval
    ) async -> TimeInterval {
        let current: RobotState?
        do { current = try await client.fullState() }
        catch { current = nil }

        var maxDelta: Double = 0
        if let h = headPose, let cur = current?.headPose {
            maxDelta = max(maxDelta, abs(h.roll - cur.roll))
            maxDelta = max(maxDelta, abs(h.pitch - cur.pitch))
            maxDelta = max(maxDelta, abs(h.yaw - cur.yaw))
        }
        if let b = bodyYaw, let curBody = current?.bodyYaw {
            maxDelta = max(maxDelta, abs(b - curBody))
        }
        if let a = antennas, let curAnt = current?.antennasPosition {
            maxDelta = max(maxDelta, abs(a.right - curAnt.right))
            maxDelta = max(maxDelta, abs(a.left - curAnt.left))
        }
        let minSafe = maxDelta / config.maxGotoVelocityRadPerS
        return max(requested, minSafe)
    }

    // MARK: - Recorded moves (shelf-safe allowlist)

    public func playRecordedMove(
        dataset: String,
        move: String,
        force: Bool = false
    ) async throws {
        if !force, !config.shelfSafeMoves.contains(move) {
            await logBus.publish(.sidecarLog(
                sidecar: "motion-guard",
                level: .warn,
                message: "blocked recorded move '\(move)' — not in shelf-safe allowlist",
                fields: ["dataset": dataset, "move": move]
            ))
            throw MotionGuardError.notShelfSafe(move: move)
        }
        try await client.playRecordedMove(dataset: dataset, move: move)
    }

    // MARK: - Pass-throughs (non-motion calls keep raw access)

    /// Read-only state passthrough. The guard doesn't need to gate
    /// reads, and callers (face tracker, state subscriber) need them
    /// at high frequency.
    public func fullState() async throws -> RobotState {
        try await client.fullState()
    }

    public func stopMove() async throws {
        try await client.stopMove()
    }

    public func setMotorMode(_ mode: MotorMode) async throws {
        try await client.setMotorMode(mode)
    }

    /// `goToSleep` is the daemon's `goto_sleep` recorded move. It's
    /// authored to land in a controlled slump, so it bypasses the
    /// shelf-safe allowlist (it IS the safe pose). Disables motors
    /// at the end.
    public func goToSleep() async throws {
        try await client.goToSleep()
    }

    /// `wakeUp` is the daemon's wake sequence: pre-seed → enable
    /// motors → 2 s minjerk goto to neutral. Same justification as
    /// `goToSleep` — controlled, intentional, by design safe.
    public func wakeUp() async throws {
        try await client.wakeUp()
    }
}

public enum MotionGuardError: Error, LocalizedError {
    case notShelfSafe(move: String)

    public var errorDescription: String? {
        switch self {
        case .notShelfSafe(let move):
            return "recorded move '\(move)' is not on the shelf-safe allowlist; use force=true to override (CHECK SHELF FIRST)"
        }
    }
}

File: OnBot/rocky_media_relay/rocky_media_relay/motion_guard.py
"""On-bot motion guard.

Mirrors the Swift `MotionGuard` (`Sources/RobotLink/MotionGuard.swift`)
on the bot, in front of the Pollen daemon. Every motion call from
Mac (or any other client) arrives at :8042/api/motion/* on this
relay; we validate against the safety constraints, then forward to
the daemon's localhost:8000/api/move/*.

This is the bot-side half of defence-in-depth motion safety. The
Mac-side `MotionGuard` catches *our* bugs early (fast failure, fast
feedback during development). This bot-side guard catches everything
else — a future SDK app on the bot, a debug curl from another
machine, a malicious request — none of them can reach the daemon
without passing through here.

Pollen's docs say the daemon clamps positions to valid ranges, but
the daemon does NOT enforce velocity, slew-rate, single-in-flight,
or any "shelf-safe" recorded-move allowlist. So those guards have to
live somewhere; this is the canonical home.

Guards enforced (same five as the Swift side):
  1. Slew-rate limit on set_target (0.05 rad / call ≈ 143°/s @ 50 Hz)
  2. Velocity clamp on goto (head + body + antennas, 1.5 rad/s default)
  3. Duration floor on goto (0.4 s)
  4. Single-in-flight goto (new awaits prior)
  5. Shelf-safe allowlist for recorded moves (force=true to override)
  6. Head-body yaw delta cap (Pollen-documented 65°)
"""

from __future__ import annotations

import asyncio
import logging
import math
import time
from dataclasses import dataclass, field
from typing import Any, Optional

import httpx

LOG = logging.getLogger("rocky_media_relay.motion_guard")

DAEMON_BASE = "http://127.0.0.1:8000"


@dataclass
class GuardConfig:
    """Tunable thresholds. Defaults match the Swift MotionGuard.Config."""

    max_set_target_slew_rad: float = 0.05         # ~143°/s at 50 Hz
    min_goto_duration_s: float = 0.4              # snap-floor
    max_goto_velocity_rad_per_s: float = 1.5      # ~86°/s
    max_head_body_yaw_delta_rad: float = 65.0 * math.pi / 180.0
    shelf_safe_moves: frozenset[str] = field(
        default_factory=lambda: frozenset({
            "amazed1", "attentive1", "calming1", "cheerful1",
            "curious1", "grateful1", "helpful1", "indifferent1",
            "inquiring1", "no1", "no_sad1", "proud1", "relief1",
            "sad1", "serenity1", "shy1", "thoughtful1", "tired1",
            "understanding1", "welcoming1", "yes1", "yes_sad1",
            "downcast1", "lonely1", "loving1", "uncertain1",
            # The wake / sleep recorded moves are intentionally
            # included — they're designed for the desk shelf.
            "wake_up", "goto_sleep",
        })
    )


class MotionGuard:
    def __init__(self, config: Optional[GuardConfig] = None) -> None:
        self.config = config or GuardConfig()
        self._http = httpx.AsyncClient(
            base_url=DAEMON_BASE, timeout=httpx.Timeout(10.0, connect=2.0)
        )
        # Slew limiter state.
        self._last_target: Optional[dict[str, Any]] = None
        # Single-in-flight goto serialisation.
        self._goto_lock = asyncio.Lock()

    async def aclose(self) -> None:
        await self._http.aclose()

    # ------------------------------------------------------------------
    # setTarget (slew-rate limited)
    # ------------------------------------------------------------------

    async def set_target(self, target: dict[str, Any]) -> httpx.Response:
        limited = self._slew_limit(target)
        self._last_target = limited
        return await self._http.post("/api/move/set_target", json=limited)

    def _slew_limit(self, target: dict[str, Any]) -> dict[str, Any]:
        prev = self._last_target
        if not prev:
            return target
        max_slew = self.config.max_set_target_slew_rad
        out = dict(target)

        new_head = target.get("target_head_pose")
        prev_head = prev.get("target_head_pose")
        if new_head is not None and prev_head is not None:
            out["target_head_pose"] = {
                "roll": _clamp_delta(new_head.get("roll", 0), prev_head.get("roll", 0), max_slew),
                "pitch": _clamp_delta(new_head.get("pitch", 0), prev_head.get("pitch", 0), max_slew),
                "yaw": _clamp_delta(new_head.get("yaw", 0), prev_head.get("yaw", 0), max_slew),
            }

        new_ant = target.get("target_antennas")
        prev_ant = prev.get("target_antennas")
        if new_ant is not None and prev_ant is not None and len(new_ant) == 2 and len(prev_ant) == 2:
            out["target_antennas"] = [
                _clamp_delta(new_ant[0], prev_ant[0], max_slew),
                _clamp_delta(new_ant[1], prev_ant[1], max_slew),
            ]

        new_body = target.get("target_body_yaw")
        prev_body = prev.get("target_body_yaw")
        if new_body is not None and prev_body is not None:
            out["target_body_yaw"] = _clamp_delta(new_body, prev_body, max_slew)

        return out

    # ------------------------------------------------------------------
    # goto (velocity clamp + duration floor + single-in-flight +
    #       head/body yaw delta gate)
    # ------------------------------------------------------------------

    async def goto(self, body: dict[str, Any]) -> httpx.Response:
        async with self._goto_lock:
            # Head-body yaw delta gate (Pollen-documented constraint).
            body = await self._reshape_for_yaw_delta(body)

            # Duration floor + velocity clamp.
            requested = max(
                float(body.get("duration", 0.7)),
                self.config.min_goto_duration_s,
            )
            safe = await self._safe_duration(body, requested)
            if safe > requested:
                LOG.warning(
                    "goto duration stretched %.2fs → %.2fs (velocity cap)",
                    requested, safe,
                )
            body = {**body, "duration": safe}
            return await self._http.post("/api/move/goto", json=body)

    async def _safe_duration(self, body: dict[str, Any], requested: float) -> float:
        try:
            cur = await self._get_current_state()
        except Exception as exc:  # noqa: BLE001
            LOG.warning("state read failed (%s); using requested duration", exc)
            return requested

        max_delta = 0.0
        head = body.get("head_pose")
        if head is not None and cur is not None:
            cur_head = cur.get("head_pose", {})
            for k in ("roll", "pitch", "yaw"):
                if k in head and k in cur_head:
                    max_delta = max(max_delta, abs(head[k] - cur_head[k]))
        body_yaw = body.get("body_yaw")
        if body_yaw is not None and cur is not None and cur.get("body_yaw") is not None:
            max_delta = max(max_delta, abs(body_yaw - cur["body_yaw"]))
        antennas = body.get("antennas")
        if antennas is not None and cur is not None:
            cur_ant = cur.get("antennas_position", [0.0, 0.0])
            if len(antennas) == 2 and len(cur_ant) == 2:
                max_delta = max(max_delta, abs(antennas[0] - cur_ant[0]))
                max_delta = max(max_delta, abs(antennas[1] - cur_ant[1]))
        min_safe = max_delta / self.config.max_goto_velocity_rad_per_s
        return max(requested, min_safe)

    async def _reshape_for_yaw_delta(self, body: dict[str, Any]) -> dict[str, Any]:
        head = body.get("head_pose")
        body_yaw = body.get("body_yaw")
        try:
            cur = await self._get_current_state()
        except Exception:  # noqa: BLE001
            cur = None
        cur_head_yaw = (cur or {}).get("head_pose", {}).get("yaw", 0.0) if cur else 0.0
        cur_body_yaw = (cur or {}).get("body_yaw", 0.0) if cur else 0.0
        target_head_yaw = head.get("yaw", cur_head_yaw) if head is not None else cur_head_yaw
        target_body_yaw = body_yaw if body_yaw is not None else cur_body_yaw
        delta = target_head_yaw - target_body_yaw
        max_delta = self.config.max_head_body_yaw_delta_rad
        if abs(delta) <= max_delta:
            return body
        excess = abs(delta) - max_delta
        direction = 1.0 if delta > 0 else -1.0
        new_head_yaw = target_head_yaw
        new_body_yaw = target_body_yaw
        if head is not None and body_yaw is not None:
            new_head_yaw -= direction * excess / 2
            new_body_yaw += direction * excess / 2
        elif head is not None:
            new_head_yaw -= direction * excess
        elif body_yaw is not None:
            new_body_yaw += direction * excess
        LOG.warning(
            "yaw-delta limit: head %.1f° + body %.1f° (Δ %.1f°) → head %.1f° + body %.1f° (Δ %.1f°)",
            math.degrees(target_head_yaw), math.degrees(target_body_yaw),
            math.degrees(delta),
            math.degrees(new_head_yaw), math.degrees(new_body_yaw),
            math.degrees(new_head_yaw - new_body_yaw),
        )
        out = dict(body)
        if head is not None:
            out["head_pose"] = {**head, "yaw": new_head_yaw}
        if body_yaw is not None:
            out["body_yaw"] = new_body_yaw
        return out

    # ------------------------------------------------------------------
    # Recorded moves (shelf-safe allowlist)
    # ------------------------------------------------------------------

    async def play_recorded_move(
        self, dataset: str, move: str, force: bool = False
    ) -> httpx.Response:
        if not force and move not in self.config.shelf_safe_moves:
            LOG.warning(
                "blocked recorded move '%s' (dataset=%s) — not in shelf-safe allowlist",
                move, dataset,
            )
            return _denied_response(
                f"recorded move '{move}' is not on the shelf-safe allowlist; "
                "use force=true to override (CHECK SHELF FIRST)"
            )
        # Daemon endpoint shape:
        #   /api/move/play/recorded-move-dataset/{dataset}/{move}   — emotion library
        #   /api/move/play/{move}                                   — built-in (wake_up, goto_sleep)
        # The dataset is None / empty for the built-ins; the relay
        # routes accordingly.
        if dataset:
            path = f"/api/move/play/recorded-move-dataset/{dataset}/{move}"
        else:
            path = f"/api/move/play/{move}"
        return await self._http.post(path)

    # ------------------------------------------------------------------
    # Pass-throughs (non-velocity-sensitive)
    # ------------------------------------------------------------------

    async def reset_slew_baseline(self, target: dict[str, Any]) -> None:
        """Drop the slew limiter's cached prev so the *next* setTarget
        is treated as the new ground-truth. Used by the wake-up
        sequence: the pre-seed setTarget MUST land exactly at the
        current physical pose (the whole point is to avoid a motor
        snap when motors enable). Without this, slew clamping
        anchors the daemon at (previous_face_tracker_target +
        max_slew) — the motors then engage to that clamped value,
        not the physical pose, which reads as an aggressive wake.
        """
        self._last_target = target

    async def wake_up(self, goto_duration_s: float = 3.5) -> dict[str, Any]:
        """Orchestrate the wake-up sequence on the bot itself, so the
        on-bot guard owns the slew baseline reset.

        Steps (matches the Swift-side composite the daemon doesn't
        provide on its own):
          1. Read current physical pose (state/full).
          2. Reset slew baseline to that pose.
          3. setTarget(currentPose) — pre-seed; with the baseline
             reset, this passes through unclamped, so the daemon's
             commanded target now matches physical.
          4. setMotorMode(enabled) — motors engage from physical →
             physical → no snap.
          5. goto(neutral, durationS=goto_duration_s, minjerk) — the
             slow swing up to looking-forward. 3.5 s by default
             (was 2 s in the Swift composite — too fast).
        """
        # 1. Read current pose.
        try:
            r = await self._http.get(
                "/api/state/full",
                params={
                    "with_head_joints": "true",
                    "with_body_yaw": "true",
                    "with_antenna_positions": "true",
                },
            )
            cur = r.json() if r.status_code == 200 else {}
        except Exception:  # noqa: BLE001
            cur = {}
        head_pose = cur.get("head_pose") or {"roll": 0.0, "pitch": 0.0, "yaw": 0.0}
        body_yaw = cur.get("body_yaw") if cur.get("body_yaw") is not None else 0.0
        ant = cur.get("antennas_position") or [0.1745, -0.1745]

        anchor_target = {
            "target_head_pose": head_pose,
            "target_body_yaw": body_yaw,
            "target_antennas": list(ant),
        }

        # 2-3. Reset baseline + pre-seed.
        await self.reset_slew_baseline(anchor_target)
        try:
            await self._http.post("/api/move/set_target", json=anchor_target)
        except Exception as exc:  # noqa: BLE001
            LOG.warning("wake_up: pre-seed setTarget failed: %s", exc)
        await asyncio.sleep(0.05)

        # 4. Engage motors. Daemon endpoint is plural / URL-based.
        try:
            await self._http.post("/api/motors/set_mode/enabled")
        except Exception as exc:  # noqa: BLE001
            LOG.warning("wake_up: setMotorMode failed: %s", exc)
        await asyncio.sleep(0.15)

        # 5. Smooth swing to neutral. Duration is configurable so
        # the user can dial it gentler if needed.
        neutral = {"roll": 0.0, "pitch": 0.0, "yaw": 0.0}
        try:
            await self._http.post(
                "/api/move/goto",
                json={
                    "head_pose": neutral,
                    "antennas": None,
                    "body_yaw": None,
                    "duration": max(goto_duration_s, 1.0),
                    "interpolation": "minjerk",
                },
            )
        except Exception as exc:  # noqa: BLE001
            LOG.warning("wake_up: goto neutral failed: %s", exc)

        # 6. Re-assert motor enable (the Swift composite did this).
        try:
            await self._http.post("/api/motors/set_mode/enabled")
        except Exception:  # noqa: BLE001
            pass

        # Slew baseline now sits at the neutral pose (subsequent
        # tracker pushes clamp from here).
        self._last_target = {
            "target_head_pose": neutral,
            "target_body_yaw": 0.0,
            "target_antennas": list(ant),
        }
        return {"ok": True, "guard": "on-bot", "goto_duration_s": max(goto_duration_s, 1.0)}

    async def set_motor_mode(self, mode: str) -> httpx.Response:
        # Daemon endpoint shape (confirmed via :8000/openapi.json):
        #   POST /api/motors/set_mode/{mode}     ← plural "motors",
        #   no body. Earlier versions of this guard used
        #   `/api/motor/mode` with a JSON body, which 404'd silently
        #   — that's why the wake-up sequence's motor-enable step
        #   was failing and the bot wasn't actually waking.
        return await self._http.post(f"/api/motors/set_mode/{mode}")

    async def stop_move(self) -> httpx.Response:
        return await self._http.post("/api/move/stop_move")

    async def daemon_proxy(self, method: str, path: str, **kwargs) -> httpx.Response:
        """Generic passthrough for state reads / other non-motion daemon calls."""
        return await self._http.request(method, path, **kwargs)

    # ------------------------------------------------------------------
    # Internals
    # ------------------------------------------------------------------

    async def _get_current_state(self) -> Optional[dict[str, Any]]:
        try:
            r = await self._http.get(
                "/api/state/full",
                params={
                    "with_head_joints": "true",
                    "with_body_yaw": "true",
                    "with_antenna_positions": "true",
                },
            )
            if r.status_code == 200:
                return r.json()
        except Exception as exc:  # noqa: BLE001
            LOG.debug("state read failed: %s", exc)
        return None


def _clamp_delta(new: float, prev: float, max_delta: float) -> float:
    delta = new - prev
    if abs(delta) <= max_delta:
        return new
    return prev + (max_delta if delta > 0 else -max_delta)


class _DummyResponse:
    """Stand-in for httpx.Response when the guard rejects a request."""

    def __init__(self, status_code: int, payload: dict[str, Any]) -> None:
        self.status_code = status_code
        self._payload = payload

    def json(self) -> dict[str, Any]:
        return self._payload

    @property
    def text(self) -> str:
        import json as _json
        return _json.dumps(self._payload)


def _denied_response(reason: str) -> Any:
    return _DummyResponse(
        status_code=403,
        payload={"ok": False, "error": reason, "guard": "motion-guard"},
    )
